JEP.md

```diff
@@ -0,0 +1,293 @@
+
+The enhancement spans two categories - general concepts applicable across all Jupyter components, and the implementation for each concept for each Jupyter component.
+
+**General concepts:**
+
+* Event schema: The schema for a given event which is respected by all the Jupyter components across the event's lifecycle.
+* Event sinks: The "backends" which handle publishing the event to its final storage location.
+
+**Jupyter components:**
+
+* Jupyter server
+* JupyterLab
+* JupyterHub
+* Jupyter Classic
+
+## Detailed Explanation
+
+### Event Schema
+
+Each event is published conforming a given event schema and this is the contract followed by each step in the event lifecycle. These are referred to by all publishers regardless of whether the event comes from the server or the browser.
+
+Example event schema
+
+```json
+{
+    "name": "org.jupyter.kernel_lifecycle_event",
+    "version": 1,
+    "title": "Event emitted when a kernel lifecycle state changes",
+    "description": "Records each invocation of any command in JupyterLab",
+    "type": "object",
+    "args": {
+        "kernel_name": {
+            "type": "string",
+            "description": "Name of the kernel"
+        },
+        "state": {
+            "type": "string",
+            "description": "The new state of the kernel"
+        }
+    }
+}
+```
+
+Schema validation is done in the core telemetry framework that routes events from publishers to sinks, so that each event sink does not need to implement schema validation.
+
+**Open Questions**:
+
+* How do organizations deploy custom schemas?
+* Where are the schemas for public events stored?
+
+### Event Sinks
+
+Event sinks are the backends where events published to.  Event sinks can be configured from the browser as well as the server. Custom sinks can be implemented by extending the interface for the given Jupyter component. 
+See [Python interface](#python-event-sink-interface) and the [JupyterLab interface](#jupyterlab-event-sink-interface)
+
+### Jupyter Server
+
+#### Python publisher library
+
+The Python event publisher library provides extension developers and other internal components an interface to record events. This is agnostic of the event sink back-end and the other deployment configurations.
+
+```python
+import jupyter.telemetry as telemetry
+
+telemetry.record_event(
+    name='org.jupyter.kernel_lifecycle_event',
+    schema_version='1',
+    args={
+        'kernel_name': 'python3',
+        'state': 'restarted'
+    }
+)
+```
+
+#### Python Event Sink Interface
+
+The event sink interface
+
+```python
+class MyCustomEventSink(JupyterEventSink):
+
+    def consume_event(event_data):
+        # Business logic to publish event into the backend.
+```
+
+#### Configuration
+
+A setting on the Jupyter server allows operators to configure various options of the event sinks such as:
+
+* The event sink implementations
+* The list of event names to whitelist or blacklist
+
+```bash
+jupyter notebook \
```

Comment:
Perfect.

Reply:
What do you think about requiring admins to *whitelist* event schemas rather than blacklist them? I think that's a fair gardrail to make our defaults to collect as little information as possible, and require admins to make informed choices on what they want to collect and why.

Reply:
@yuvipanda whitelist makes sense to me.

Reply:
I had both "blacklist" and "whitelist" in the document, but the example only has blacklist. IF we only offer one, I agree we should go for whitelist.

What do folks think about offering both options? (blacklist always take preference over whitelist in this world)